### ISSUE TITLE
use pytorch and torch in R in same project?

### ISSUE THREAD

**Issue by egillax** [mlverse/torch#815]
Is it possible to use pytorch through reticulate and torch in R in the same project? 

If I try to load one after the other has been loaded I get errors:

```R
> reticulate::use_python('/home/egill/projects/transfer_learning/venv/bin/python')
> pytorch <- reticulate::import('torch')
> library(torch)

Warning message:
Torch failed to start, restart your R session to try again. /home/egill/R/x86_64-pc-linux-gnu-library/4.1/torch/lib/liblantern.so - /home/egill/R/x86_64-pc-linux-gnu-library/4.1/torch/lib/liblantern.so: undefined symbol: _ZNK5torch8autograd4Node4nameB5cxx11Ev 
```
If its the other way around:

```R
> library(torch)
> reticulate::use_python('/home/egill/projects/transfer_learning/venv/bin/python')
> pytorch <- reticulate::import('torch')
Error in py_module_import(module, convert = convert) : 
  ImportError: /home/egill/projects/transfer_learning/venv/lib/python3.9/site-packages/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch8autograd8profiler14ProfilerResult4saveERKSs
```
Is this at all possible?

**Comment by dfalbel** [comment]
In theory this is possible if the R version and the python version are using the exact same version of libtorch. Currently, the CRAN version of torch uses libtorch 1.10.2, so you should be able to use them together.

``` r
library(torch)
torch_randn(1)
#> torch_tensor
#>  2.8685
#> [ CPUFloatType{1} ]

pytorch <- reticulate::import("torch")
pytorch$randn(1L)
#> tensor([0.4279])
```

<sup>Created on 2022-05-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

You can install specified versions of pip libraries with: 

```
reticulate::py_install("torch==1.10.2", pip = TRUE)
```

**Comment by egillax** [comment]
No matter what I do I can't seem to get this to work on ubuntu. I've made sure both pytorch and torch are 1.10.2 and use cuda 11.3. I've reinstalled both many times.

Any idea how I can troubleshoot this?

**Comment by dfalbel** [comment]
Maybe you need to install pytorch 1.10.2 for cuda 11.3? Something like:

```
pip3 install torch==1.10.2 --extra-index-url https://download.pytorch.org/whl/cu113
```

**Comment by egillax** [comment]
After some digging I did figure this out. This has to do with if the underlying libtorch is the cxx11 abi version. The official pytorch version both using conda or pip to install is the non-abi version. While the one supplied with Torch in R is the abi version. These need to match if you're using Linux.

The solution for me was to install pytorch using conda and the conda-forge channel, since apparently those binaries are abi-ones. 

```conda install  pytorch=1.10.2=cuda111py38hc64aeea_1 -c conda-forge``` was the command I used for cuda11.1 (conda-forge doesn't seem to have cuda11.3)
